File: internal/changelog/changelog_test.go
package changelog

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

const fixtureChangelog = `# Changelog

Todas as mudanças notáveis deste projeto serão documentadas neste arquivo.

## [Unreleased]

### Added
- feature em desenvolvimento

## [6.10.0] - 2026-08-14

### Added
- comando trackfw changelog

### Fixed
- bug no parser

## [6.9.1] - 2026-08-01

### Fixed
- correção pontual
`

func TestParseSectionsThreeSections(t *testing.T) {
	sections, err := ParseSections(fixtureChangelog)
	if err != nil {
		t.Fatalf("ParseSections: erro inesperado: %v", err)
	}
	if len(sections) != 3 {
		t.Fatalf("esperava 3 seções, obteve %d: %+v", len(sections), sections)
	}

	unreleased := sections[0]
	if unreleased.Version != "Unreleased" {
		t.Errorf("sections[0].Version = %q, esperava %q", unreleased.Version, "Unreleased")
	}
	if unreleased.Date != "" {
		t.Errorf("sections[0].Date = %q, esperava vazio", unreleased.Date)
	}
	if !strings.Contains(unreleased.Body, "feature em desenvolvimento") {
		t.Errorf("sections[0].Body não contém o conteúdo esperado: %q", unreleased.Body)
	}
	if strings.Contains(unreleased.Body, "## [") {
		t.Errorf("sections[0].Body não deve conter cabeçalho de outra seção: %q", unreleased.Body)
	}

	v6100 := sections[1]
	if v6100.Version != "6.10.0" {
		t.Errorf("sections[1].Version = %q, esperava %q", v6100.Version, "6.10.0")
	}
	if v6100.Date != "2026-08-14" {
		t.Errorf("sections[1].Date = %q, esperava %q", v6100.Date, "2026-08-14")
	}
	if !strings.Contains(v6100.Body, "comando trackfw changelog") || !strings.Contains(v6100.Body, "bug no parser") {
		t.Errorf("sections[1].Body não contém o conteúdo esperado: %q", v6100.Body)
	}

	v691 := sections[2]
	if v691.Version != "6.9.1" {
		t.Errorf("sections[2].Version = %q, esperava %q", v691.Version, "6.9.1")
	}
	if v691.Date != "2026-08-01" {
		t.Errorf("sections[2].Date = %q, esperava %q", v691.Date, "2026-08-01")
	}
	if !strings.Contains(v691.Body, "correção pontual") {
		t.Errorf("sections[2].Body não contém o conteúdo esperado: %q", v691.Body)
	}
}

func TestFirstSectionEmptyList(t *testing.T) {
	_, err := FirstSection(nil)
	if err == nil {
		t.Fatal("FirstSection(nil): esperava erro, obteve nil")
	}
	if err.Error() != "CHANGELOG.md has no version sections" {
		t.Errorf("FirstSection(nil): mensagem de erro = %q, esperava %q", err.Error(), "CHANGELOG.md has no version sections")
	}
}

func TestFirstSectionReturnsFirst(t *testing.T) {
	sections, err := ParseSections(fixtureChangelog)
	if err != nil {
		t.Fatalf("ParseSections: erro inesperado: %v", err)
	}
	first, err := FirstSection(sections)
	if err != nil {
		t.Fatalf("FirstSection: erro inesperado: %v", err)
	}
	if first.Version != "Unreleased" {
		t.Errorf("FirstSection.Version = %q, esperava %q", first.Version, "Unreleased")
	}
}

func TestFindVersionExisting(t *testing.T) {
	sections, err := ParseSections(fixtureChangelog)
	if err != nil {
		t.Fatalf("ParseSections: erro inesperado: %v", err)
	}
	got, err := FindVersion(sections, "6.10.0")
	if err != nil {
		t.Fatalf("FindVersion(6.10.0): erro inesperado: %v", err)
	}
	if got.Version != "6.10.0" {
		t.Errorf("FindVersion(6.10.0).Version = %q, esperava %q", got.Version, "6.10.0")
	}
}

func TestFindVersionMissing(t *testing.T) {
	sections, err := ParseSections(fixtureChangelog)
	if err != nil {
		t.Fatalf("ParseSections: erro inesperado: %v", err)
	}
	_, err = FindVersion(sections, "999.0.0")
	if err == nil {
		t.Fatal("FindVersion(999.0.0): esperava erro, obteve nil")
	}
	want := `version "999.0.0" not found in CHANGELOG.md`
	if err.Error() != want {
		t.Errorf("FindVersion(999.0.0): mensagem de erro = %q, esperava %q", err.Error(), want)
	}
}

func TestFindVersionNormalizesVPrefix(t *testing.T) {
	sections, err := ParseSections(fixtureChangelog)
	if err != nil {
		t.Fatalf("ParseSections: erro inesperado: %v", err)
	}
	withV, err := FindVersion(sections, "v6.10.0")
	if err != nil {
		t.Fatalf("FindVersion(v6.10.0): erro inesperado: %v", err)
	}
	withoutV, err := FindVersion(sections, "6.10.0")
	if err != nil {
		t.Fatalf("FindVersion(6.10.0): erro inesperado: %v", err)
	}
	if withV.Version != withoutV.Version || withV.Body != withoutV.Body {
		t.Errorf("FindVersion(v6.10.0) != FindVersion(6.10.0): %+v vs %+v", withV, withoutV)
	}
}

func TestFormatSectionWithDate(t *testing.T) {
	s := Section{Version: "6.10.0", Date: "2026-08-14", Body: "### Added\n- x\n"}
	got := FormatSection(s)
	want := "## [6.10.0] - 2026-08-14\n\n### Added\n- x\n"
	if got != want {
		t.Errorf("FormatSection = %q, esperava %q", got, want)
	}
}

func TestFormatSectionWithoutDate(t *testing.T) {
	s := Section{Version: "Unreleased", Date: "", Body: "### Added\n- x\n"}
	got := FormatSection(s)
	want := "## [Unreleased]\n\n### Added\n- x\n"
	if got != want {
		t.Errorf("FormatSection = %q, esperava %q", got, want)
	}
}

func TestFormatSectionDoesNotDuplicateBlankLineWhenBodyStartsWithNewline(t *testing.T) {
	// Body como vem de ParseSections para uma seção cujo cabeçalho é seguido
	// de linha em branco no CHANGELOG.md original (caso real deste repo):
	// a primeira linha do Body já é vazia.
	s := Section{Version: "Unreleased", Date: "", Body: "\n### Added\n- x\n"}
	got := FormatSection(s)
	want := "## [Unreleased]\n\n### Added\n- x\n"
	if got != want {
		t.Errorf("FormatSection = %q, esperava %q (sem linha em branco duplicada)", got, want)
	}
}

func TestReadMissingFile(t *testing.T) {
	dir := t.TempDir()
	_, err := Read(dir)
	if err == nil {
		t.Fatal("Read: esperava erro, obteve nil")
	}
	if err.Error() != "CHANGELOG.md not found — nothing to show" {
		t.Errorf("Read: mensagem de erro = %q, esperava %q", err.Error(), "CHANGELOG.md not found — nothing to show")
	}
}

func TestReadExistingFile(t *testing.T) {
	dir := t.TempDir()
	if err := os.WriteFile(filepath.Join(dir, "CHANGELOG.md"), []byte(fixtureChangelog), 0o644); err != nil {
		t.Fatalf("setup: %v", err)
	}
	got, err := Read(dir)
	if err != nil {
		t.Fatalf("Read: erro inesperado: %v", err)
	}
	if got != fixtureChangelog {
		t.Errorf("Read: conteúdo divergente do fixture")
	}
}

File: npm/tests/changelog.test.js
'use strict'

const test = require('node:test')
const assert = require('node:assert/strict')
const fs = require('node:fs')
const os = require('node:os')
const path = require('node:path')

const changelog = require('../src/changelog')

const fixtureChangelog = `# Changelog

Todas as mudanças notáveis deste projeto serão documentadas neste arquivo.

## [Unreleased]

### Added
- feature em desenvolvimento

## [6.10.0] - 2026-08-14

### Added
- comando trackfw changelog

### Fixed
- bug no parser

## [6.9.1] - 2026-08-01

### Fixed
- correção pontual
`

// --- parseSections -----------------------------------------------------

test('parseSections separa 3 seções (Unreleased + 2 versões)', () => {
  const sections = changelog.parseSections(fixtureChangelog)
  assert.equal(sections.length, 3)

  const unreleased = sections[0]
  assert.equal(unreleased.version, 'Unreleased')
  assert.equal(unreleased.date, '')
  assert.ok(unreleased.body.includes('feature em desenvolvimento'))
  assert.ok(!unreleased.body.includes('## ['))

  const v6100 = sections[1]
  assert.equal(v6100.version, '6.10.0')
  assert.equal(v6100.date, '2026-08-14')
  assert.ok(v6100.body.includes('comando trackfw changelog'))
  assert.ok(v6100.body.includes('bug no parser'))

  const v691 = sections[2]
  assert.equal(v691.version, '6.9.1')
  assert.equal(v691.date, '2026-08-01')
  assert.ok(v691.body.includes('correção pontual'))
})

// --- firstSection --------------------------------------------------------

test('firstSection lança erro para lista vazia', () => {
  assert.throws(() => changelog.firstSection([]), /CHANGELOG\.md has no version sections/)
})

test('firstSection retorna a primeira seção', () => {
  const sections = changelog.parseSections(fixtureChangelog)
  const first = changelog.firstSection(sections)
  assert.equal(first.version, 'Unreleased')
})

// --- findVersion -----------------------------------------------------------

test('findVersion encontra versão existente', () => {
  const sections = changelog.parseSections(fixtureChangelog)
  const got = changelog.findVersion(sections, '6.10.0')
  assert.equal(got.version, '6.10.0')
})

test('findVersion lança erro para versão ausente', () => {
  const sections = changelog.parseSections(fixtureChangelog)
  assert.throws(
    () => changelog.findVersion(sections, '999.0.0'),
    /version "999\.0\.0" not found in CHANGELOG\.md/
  )
})

test('findVersion normaliza prefixo "v"', () => {
  const sections = changelog.parseSections(fixtureChangelog)
  const withV = changelog.findVersion(sections, 'v6.10.0')
  const withoutV = changelog.findVersion(sections, '6.10.0')
  assert.equal(withV.version, withoutV.version)
  assert.equal(withV.body, withoutV.body)
})

// --- formatSection -----------------------------------------------------

test('formatSection com data', () => {
  const s = { version: '6.10.0', date: '2026-08-14', body: '### Added\n- x\n' }
  const got = changelog.formatSection(s)
  assert.equal(got, '## [6.10.0] - 2026-08-14\n\n### Added\n- x\n')
})

test('formatSection sem data', () => {
  const s = { version: 'Unreleased', date: '', body: '### Added\n- x\n' }
  const got = changelog.formatSection(s)
  assert.equal(got, '## [Unreleased]\n\n### Added\n- x\n')
})

test('formatSection não duplica linha em branco quando body começa com \\n', () => {
  // Body como vem de parseSections para uma seção cujo cabeçalho é seguido
  // de linha em branco no CHANGELOG.md original (caso real deste repo): a
  // primeira linha do body já é vazia.
  const s = { version: 'Unreleased', date: '', body: '\n### Added\n- x\n' }
  const got = changelog.formatSection(s)
  assert.equal(got, '## [Unreleased]\n\n### Added\n- x\n')
})

// --- read ----------------------------------------------------------------

test('read lança erro quando CHANGELOG.md não existe', () => {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-changelog-'))
  assert.throws(() => changelog.read(dir), /CHANGELOG\.md not found — nothing to show/)
})

test('read retorna conteúdo do arquivo existente', () => {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-changelog-'))
  fs.writeFileSync(path.join(dir, 'CHANGELOG.md'), fixtureChangelog)
  const got = changelog.read(dir)
  assert.equal(got, fixtureChangelog)
})

// --- comando `trackfw changelog` ------------------------------------------

// commander's Command instance carries option state across repeated
// parseAsync() calls on the same object -- require a fresh module instance
// per test to avoid cross-test leakage (same pattern as npm/tests/adr.test.js).
function freshChangelogCmd() {
  const resolved = require.resolve('../src/commands/changelog')
  delete require.cache[resolved]
  return require('../src/commands/changelog')
}

async function captureStdoutAsync(fn) {
  const orig = process.stdout.write.bind(process.stdout)
  let out = ''
  process.stdout.write = (chunk) => {
    out += chunk
    return true
  }
  try {
    await fn()
  } finally {
    process.stdout.write = orig
  }
  return out
}

let __origCwd
test.beforeEach(() => {
  __origCwd = process.cwd()
})
test.afterEach(() => {
  process.chdir(__origCwd)
  process.exitCode = 0
})

function chdirToProjectWithChangelog(content) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-changelog-cmd-'))
  if (content !== '') {
    fs.writeFileSync(path.join(dir, 'CHANGELOG.md'), content)
  }
  process.chdir(dir)
  return dir
}

test('trackfw changelog sem flags imprime a primeira seção', async () => {
  chdirToProjectWithChangelog(fixtureChangelog)
  const cmd = freshChangelogCmd()
  const out = await captureStdoutAsync(() => cmd.parseAsync(['node', 'changelog']))
  assert.equal(out, '## [Unreleased]\n\n### Added\n- feature em desenvolvimento\n')
})

test('trackfw changelog --version 6.10.0 imprime a seção da versão', async () => {
  chdirToProjectWithChangelog(fixtureChangelog)
  const cmd = freshChangelogCmd()
  const out = await captureStdoutAsync(() => cmd.parseAsync(['node', 'changelog', '--version', '6.10.0']))
  assert.equal(
    out,
    '## [6.10.0] - 2026-08-14\n\n### Added\n- comando trackfw changelog\n\n### Fixed\n- bug no parser\n'
  )
})

async function captureStderrAsync(fn) {
  const origError = console.error
  let out = ''
  console.error = (msg) => {
    out += msg + '\n'
  }
  try {
    await fn()
  } finally {
    console.error = origError
  }
  return out
}

test('trackfw changelog --version inexistente reporta erro e exitCode 1', async () => {
  chdirToProjectWithChangelog(fixtureChangelog)
  const cmd = freshChangelogCmd()
  const errOut = await captureStderrAsync(() => cmd.parseAsync(['node', 'changelog', '--version', '999.0.0']))
  assert.match(errOut, /Error: version "999\.0\.0" not found in CHANGELOG\.md/)
  assert.equal(process.exitCode, 1)
})

test('trackfw changelog --all imprime o arquivo inteiro', async () => {
  chdirToProjectWithChangelog(fixtureChangelog)
  const cmd = freshChangelogCmd()
  const out = await captureStdoutAsync(() => cmd.parseAsync(['node', 'changelog', '--all']))
  assert.equal(out, fixtureChangelog)
})

test('trackfw changelog sem CHANGELOG.md reporta erro claro e exitCode 1', async () => {
  chdirToProjectWithChangelog('')
  const cmd = freshChangelogCmd()
  const errOut = await captureStderrAsync(() => cmd.parseAsync(['node', 'changelog']))
  assert.match(errOut, /Error: CHANGELOG\.md not found — nothing to show/)
  assert.equal(process.exitCode, 1)
})
